src/mono/mono/mini/interp/interp-simd.c

```diff
@@ -0,0 +1,605 @@
+interp_v128_i4_op_subtraction (gpointer res, gpointer v1, gpointer v2)
+{
+	*(v128_i4*)res = *(v128_i4*)v1 - *(v128_i4*)v2;
+}
+
+// op_BitwiseAnd
+static void
+interp_v128_op_bitwise_and (gpointer res, gpointer v1, gpointer v2)
+{
+	*(v128_i4*)res = *(v128_i4*)v1 & *(v128_i4*)v2;
+}
+
+// op_BitwiseOr
+static void
+interp_v128_op_bitwise_or (gpointer res, gpointer v1, gpointer v2)
+{
+	*(v128_i4*)res = *(v128_i4*)v1 | *(v128_i4*)v2;
+}
+
+// op_Division
+static void
+interp_v128_i1_op_division (gpointer res, gpointer v1, gpointer v2)
+{
+	*(v128_i1*)res = *(v128_i1*)v1 / *(v128_i1*)v2;
+}
+
+static void
+interp_v128_u1_op_division (gpointer res, gpointer v1, gpointer v2)
+{
+	*(v128_u1*)res = *(v128_u1*)v1 / *(v128_u1*)v2;
+}
+
+static void
+interp_v128_i2_op_division (gpointer res, gpointer v1, gpointer v2)
+{
+	*(v128_i2*)res = *(v128_i2*)v1 / *(v128_i2*)v2;
+}
+
+static void
+interp_v128_u2_op_division (gpointer res, gpointer v1, gpointer v2)
+{
+	*(v128_u2*)res = *(v128_u2*)v1 / *(v128_u2*)v2;
+}
+
+static void
+interp_v128_i4_op_division (gpointer res, gpointer v1, gpointer v2)
+{
+	*(v128_i4*)res = *(v128_i4*)v1 / *(v128_i4*)v2;
+}
+
+static void
+interp_v128_u4_op_division (gpointer res, gpointer v1, gpointer v2)
+{
+	*(v128_u4*)res = *(v128_u4*)v1 / *(v128_u4*)v2;
+}
+
+// op_Equality
+static void
+interp_v128_op_equality (gpointer res, gpointer v1, gpointer v2)
+{
+	gint64 *v1_cast = (gint64*)v1;
+	gint64 *v2_cast = (gint64*)v2;
+
+	if (*v1_cast == *v2_cast && *(v1_cast + 1) == *(v2_cast + 1))
+		*(gint32*)res = 1;
+	else
+		*(gint32*)res = 0;
+}
+
+// op_ExclusiveOr
+static void
+interp_v128_op_exclusive_or (gpointer res, gpointer v1, gpointer v2)
+{
+	*(v128_i4*)res = *(v128_i4*)v1 ^ *(v128_i4*)v2;
+}
+
+// op_Inequality
+static void
+interp_v128_op_inequality (gpointer res, gpointer v1, gpointer v2)
+{
+	gint64 *v1_cast = (gint64*)v1;
+	gint64 *v2_cast = (gint64*)v2;
+
+	if (*v1_cast == *v2_cast && *(v1_cast + 1) == *(v2_cast + 1))
+		*(gint32*)res = 0;
+	else
+		*(gint32*)res = 1;
+}
+
+// op_Addition
+static void
+interp_v128_i1_op_multiply (gpointer res, gpointer v1, gpointer v2)
+{
+	*(v128_i1*)res = *(v128_i1*)v1 * *(v128_i1*)v2;
+}
+
+static void
+interp_v128_i2_op_multiply (gpointer res, gpointer v1, gpointer v2)
+{
+	*(v128_i2*)res = *(v128_i2*)v1 * *(v128_i2*)v2;
+}
+
+static void
+interp_v128_i4_op_multiply (gpointer res, gpointer v1, gpointer v2)
+{
+	*(v128_i4*)res = *(v128_i4*)v1 * *(v128_i4*)v2;
+}
+
+// op_UnaryNegation
+static void
+interp_v128_i1_op_negation (gpointer res, gpointer v1)
+{
+	*(v128_i1*)res = - (*(v128_i1*)v1);
+}
+
+static void
+interp_v128_i2_op_negation (gpointer res, gpointer v1)
+{
+	*(v128_i2*)res = - (*(v128_i2*)v1);
+}
+
+static void
+interp_v128_i4_op_negation (gpointer res, gpointer v1)
+{
+	*(v128_i4*)res = - (*(v128_i4*)v1);
+}
+
+// op_LeftShift
+static void
+interp_v128_i1_op_left_shift (gpointer res, gpointer v1, gpointer s1)
+{
+	*(v128_i1*)res = *(v128_i1*)v1 << *(gint32*)s1;
```

Comment:
C's behavior around shifting doesn't necessarily match that of C#'s or the underlying platform's.

Might be worth validating this has the expected behavior for each `T`.

Reply:
Are you referring to shifting of negative numbers ? In mono interpreter we actually use the C shifting operator and haven't encountered any problem on any platform so far. I also tested locally this vectorized code for all integer types and I found no issues.

Reply:
C# and by extension most shifting APIs exposed by .NET explicitly mask the input bits. That is, C# strictly defines shifting for `int` as `x << (y & 31)`. It likewise defines it for `long` as `x << (y & 63)`. For small types, it implements it in terms of `int` and then downcasts since it doesn't typically have "proper" support for small types.

Thus `x << 33` is treated as `x << 1` and does not return `0`. Likewise, `x << -1` is treated as `x << 31`, it is not treated as `x >> 1`.

x86/x64 define the shifting behavior as masking as well, so this tends to "just work" for `int`/`long`. But, in C/C++ it is strictly "undefined behavior" and so MSVC/GCC/Clang compilers do not have to match the underlying platform or be consistent with eachother, can make differing optimizations when constant folding or based on range propagation, etc.

Being explicit in these operations helps ensure there isn't any weirdness or inconsistency with the overshifting or negative shifting case.

Additionally, since `Vector###<T>` and generic math have "proper" support for small types, we don't implement them in terms of int and thus `Vector128<byte>.ShiftLeft` is defined as `x << (y & 7)` per element.

Reply:
I'm quite confused. From what I understand from you, the shifting in c# might by different from the shifting done by the vectorized assembly operations behind the scenes. So I wrote a testcase to verify shifting for every single possible sbyte value. The problem is that, for negative and overflowing shift amounts, this test fails on all configurations, including on mono interpreter without this PR as well as on CoreCLR (I tested 7.0 rc1). I hope I didn't do anything stupid in this testcase, but for normal shifts between 0 and 7, the test passes just fine.

```
using System;
using System.Runtime.Intrinsics;

public class Program {

	public static bool CheckValue (Vector128<sbyte> vec, sbyte[] values, int index, int shift_amount)
	{
		sbyte csharp_value = (sbyte)(values [index] << shift_amount);
		sbyte vec_value = vec.GetElement (index);
		if (csharp_value != vec_value) {
			Console.WriteLine ("Error shifting {0} by {1}, ({2} != {3})", values [index], shift_amount, vec_value, csharp_value);
			return false;
		} else {
			return true;
		}
	}

	public static void TestShift ()
	{
		int num_elements = Vector128<sbyte>.Count;
		sbyte[] values = new sbyte[num_elements];

		for (int shift_amount = -9; shift_amount < 9; shift_amount++) {
			Console.WriteLine ("Testing shift {0}", shift_amount);
			bool pass = true;
			for (sbyte i = sbyte.MinValue; i < sbyte.MaxValue && pass; i++) {
				for (int k = 0; k < num_elements; k++)
					values [k] = i;

				Vector128<sbyte> vector = Vector128.Create (values);
				vector = Vector128.ShiftLeft (vector, shift_amount);

				for (int k = 0; k < num_elements && pass; k++)
					pass &= CheckValue (vector, values, k, shift_amount);
			}
			if (pass)
				Console.WriteLine ("OK");
			else
				Console.WriteLine ("NOK");
		}
	}

	public static void Main (string[] args)
	{
		TestShift ();
	}
}
```

Reply:
>  From what I understand from you, the shifting in c# might by different from the shifting done by the vectorized assembly operations behind the scenes.

We have two different scenarios to consider.

First we have the platform specific hardware intrinsics such as `System.Runtime.Intrinsics.X86.Sse2.ShiftLeftLogical` or `System.Runtime.Intrinsics.Arm.AdvSimd.ShiftLeftLogical`. These intrinsics emit exactly 1 instruction and may do different operations per platform.

We then also have cross platform hardware intrinsics such as `System.Runtime.Intrinsics.Vector128.ShiftLeft` and `System.Runtime.Intrinsics.Vector128<T>.operator <<`. These intrinsics ensure consistent and deterministic behavior across all platforms and ensure that edge cases are handled where applicable.

For the cross platform hardware intrinsics, you can see the software fallback implementation of `operator <<` [here](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector64_1.cs#L294-L305). You'll see that it calls [Scalar<T>.ShiftLeft](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Scalar.cs#L796-L854) which in turn does the shifting in C# and therefore ensures that all the shifting is "well-defined" (it masks `shiftCount` against `7`, `15`, or relies on the implicit masking C# emits for `31` and `63`).

We then also have logic for the intrinsic paths to ensure the IR is imported correctly for non-constants and that it gives us the same ultimate behavior of masking the `shiftCount`: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/gentree.cpp#L19552-L19624

> So I wrote a testcase to verify shifting for every single possible sbyte value.

I explained this bit above, but it might have gotten lost. C# (and .NET IL in general) doesn't have "proper" support for small-types (`byte`, `sbyte`, `short`, `ushort`, `bool`, or `char`). They are all implemented in terms of an implicit upcast to `int` and so for `sbyte`, your `(sbyte)(values [index] << shift_amount)` is being emitted in IL as something more like: `(sbyte)((int)(values[index]) << (shift_amount & 31))`

The intrinsics all operate directly with small types and thus have "proper support" for them. The same applies to `generic math` where `byte op byte` produces `byte`, not `int`. Thus, we ensure there remains well-defined logic in the face of small types for what we expose.

Reply:
I see. Thanks for the detailed explanation. The behavior was indeed different with or without this PR for those shift amounts. I added the masking within those intrinsics and the outputs seem to be correct now.